File: examples/kotlin/examples/ClientChatModels.kt
package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.github.cdimascio.dotenv.dotenv
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Chat Models Example - Using ekoDB Kotlin client library
 *
 * This example demonstrates Chat Models API:
 * - getChatModels: List all available chat models by provider
 * - getChatModel: Get models for a specific provider
 */
fun main() = runBlocking {
    val dotenv = dotenv()
    val baseUrl = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    val apiKey = dotenv["API_BASE_KEY"] ?: "a-test-api-key-from-ekodb"

    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()

    println("=== ekoDB Kotlin Client - Chat Models Example ===\n")

    try {
        // Example 1: List all available chat models by provider
        println("=== List Chat Models ===")
        try {
            val models = client.getChatModels()
            println("Available chat models by provider:")

            // Iterate over providers
            // `providers` is a status object, not a model list — printed below.
            for ((provider, modelList) in models.filterKeys { it != "providers" }) {
                val modelArray = modelList as? JsonArray
                if (modelArray != null && modelArray.isNotEmpty()) {
                    println("  $provider:")
                    for (model in modelArray) {
                        println("    - ${model.jsonPrimitive.content}")
                    }
                }
            }
            // Why each list looks the way it does: a rejected key reports
            // "auth_failed", a missing one "not_configured".
            println("Provider status:")
            models["providers"]?.jsonObject?.forEach { (provider, status) ->
                val row = status.jsonObject
                val detail = row["model_count"]?.jsonPrimitive?.content?.let { "$it models" }
                    ?: row["message"]?.jsonPrimitive?.content.orEmpty()
                val verified = if (row["verified"]?.jsonPrimitive?.content == "true") "" else " (unverified)"
                println("  $provider: ${row["status"]?.jsonPrimitive?.content}$verified $detail")
            }
        } catch (e: Exception) {
            println("GetChatModels error: ${e.message}")
        }
        println()

        // Example 2: Get models for a specific provider (OpenAI)
        println("=== Get OpenAI Models ===")
        try {
            val openaiModels = client.getChatModel("openai")
            val modelNames = openaiModels.map { it.jsonPrimitive.content }
            println("OpenAI models: ${modelNames.joinToString(", ").ifEmpty { "none configured" }}")
        } catch (e: Exception) {
            println("GetChatModel error: ${e.message}")
        }
        println()

        // Example 3: Get Anthropic models
        println("=== Get Anthropic Models ===")
        try {
            val anthropicModels = client.getChatModel("anthropic")
            val modelNames = anthropicModels.map { it.jsonPrimitive.content }
            println("Anthropic models: ${modelNames.joinToString(", ").ifEmpty { "none configured" }}")
        } catch (e: Exception) {
            println("GetChatModel error: ${e.message}")
        }
        println()

        // Example 4: Handle non-existent provider
        println("=== Get Non-Existent Provider ===")
        try {
            client.getChatModel("nonexistent_provider_xyz")
            println("Unexpectedly found non-existent provider")
        } catch (e: Exception) {
            println("Expected error for non-existent provider: ${e.message}")
        }

    } finally {
        client.close()
        println("\n=== Chat Models Example Complete ===")
    }
}

File: examples/rust/examples/client_chat_models.rs
//! Chat Models API Example - Using ekodb_client library
//!
//! This example demonstrates how to retrieve available chat models
//! from the ekoDB server organized by provider.

use ekodb_client::Client;
use std::env;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    // Create ekoDB client
    let client = Client::builder()
        .base_url(&base_url)
        .api_key(&api_key)
        .build()?;

    println!("✓ Client created");

    // Example 1: Get All Chat Models
    println!("\n=== Get All Chat Models ===");
    match client.get_chat_models().await {
        Ok(models) => {
            println!("Available chat models by provider:");
            println!("\nOpenAI models ({}):", models.openai.len());
            for model in &models.openai {
                println!("  - {}", model);
            }
            println!("\nAnthropic models ({}):", models.anthropic.len());
            for model in &models.anthropic {
                println!("  - {}", model);
            }
            println!("\nPerplexity models ({}):", models.perplexity.len());
            for model in &models.perplexity {
                println!("  - {}", model);
            }
            println!("\nGemini models ({}):", models.gemini.len());
            for model in &models.gemini {
                println!("  - {}", model);
            }
            // Why each list looks the way it does: a rejected key reports
            // `auth_failed`, a missing one `not_configured`.
            println!("\nProvider status:");
            for (provider, status) in &models.providers {
                let detail = match status.model_count {
                    Some(count) => format!("{count} models"),
                    None => status.message.clone().unwrap_or_default(),
                };
                println!(
                    "  {provider}: {:?}{} {detail}",
                    status.status,
                    if status.verified { "" } else { " (unverified)" }
                );
            }
        }
        Err(e) => println!("GetChatModels error: {}", e),
    }

    // Example 2: Get Models for Specific Provider
    println!("\n=== Get Models for Specific Provider ===");
    let providers = ["openai", "anthropic", "perplexity"];

    for provider in providers {
        match client.get_chat_model(provider).await {
            Ok(models) => {
                println!("\n{} models ({}):", provider, models.len());
                for model in models.iter().take(5) {
                    println!("  - {}", model);
                }
                if models.len() > 5 {
                    println!("  ... and {} more", models.len() - 5);
                }
            }
            Err(e) => println!("GetChatModel({}) error: {}", provider, e),
        }
    }

    println!("\n✓ Chat Models API example complete");

    Ok(())
}
